### PR TITLE
[REVIEW] Port nvtext create_ngrams function

### PR DESCRIPTION
**Pull request by davidwendt**
Reference issue #3965 
This function creates a column of ngrams from an existing column of strings.

### CODE REVIEW

File: cpp/src/text/generate_ngrams.cu [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    CUDF_EXPECTS( strings_count >= ngrams , "Insufficient number of strings to generate ngrams");
+    // create a temporary column view from the non-empty offsets and chars column views
+    cudf::column_view strings_view(cudf::data_type{cudf::STRING}, strings_count, 
+                                   nullptr, nullptr, 0, 0,
+                                   {non_empty_offsets_column->view(), strings.chars()});
+    strings_column = cudf::column_device_view::create(strings_view,stream);
+    d_strings = *strings_column;
+
+    // compute the number of strings of ngrams
+    auto ngrams_count = strings_count - ngrams + 1;
```
Comment (codereport) [inline cpp/src/text/generate_ngrams.cu]:
```suggestion
    auto const ngrams_count = strings_count - ngrams + 1;
```

File: cpp/src/text/generate_ngrams.cu [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                        return false;
+                    return !d_strings.element<cudf::string_view>(idx).empty();
+                }, mr, stream )->release();
+        strings_count = table_offsets.front()->size()-1;
+        return std::move(table_offsets.front());
+    } (); // this allows freeint the temporary table_offsets
+
+    if( strings_count==0 ) // if no non-empty strings, return an empty column
+        return cudf::make_empty_column(cudf::data_type{cudf::STRING});
+    CUDF_EXPECTS( strings_count >= ngrams , "Insufficient number of strings to generate ngrams");
```
Comment (vuule) [inline cpp/src/text/generate_ngrams.cu]: Should strings_count == 0 and (for example) strings_count==1 be treated differently? AFAICT they both represent the same case - insufficient number of elements. 
Either strings_count < ngrams should yield an empty column, or empty input should cause an exception.

File: cpp/src/text/generate_ngrams.cu [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    auto offsets_transformer_itr = thrust::make_transform_iterator( thrust::make_counting_iterator<cudf::size_type>(0),
+        ngram_generator_fn{d_strings,ngrams,d_separator} );
+    auto offsets_column = cudf::strings::detail::make_offsets_child_column(offsets_transformer_itr,
+                                       offsets_transformer_itr + ngrams_count,
+                                       mr, stream);
+    auto d_offsets = offsets_column->view().data<int32_t>();
+
+    // build the chars column
+    // generate the ngrams from the input strings and copy them into the chars data buffer
+    cudf::size_type total_bytes = thrust::device_pointer_cast(d_offsets)[ngrams_count];
```
Comment (vuule) [inline cpp/src/text/generate_ngrams.cu]:
```suggestion
    cudf::size_type const total_bytes = thrust::device_pointer_cast(d_offsets)[ngrams_count];
```

File: cpp/src/text/generate_ngrams.cu [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    auto chars_column = cudf::strings::detail::create_chars_child_column( ngrams_count, 0, total_bytes, mr, stream );
+    auto d_chars = chars_column->mutable_view().data<char>();
```
Comment (vuule) [inline cpp/src/text/generate_ngrams.cu]: I think this one too
```suggestion
    auto const d_chars = chars_column->mutable_view().data<char>();
```

File: cpp/src/text/generate_ngrams.cu [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+} // namespace
+
+std::unique_ptr<cudf::column> generate_ngrams( cudf::strings_column_view const& strings,
+                                               cudf::size_type ngrams = 2,
+                                               cudf::string_scalar const& separator = cudf::string_scalar{"_"},
+                                               rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                               cudaStream_t stream = 0 )
+{
+    CUDF_EXPECTS( separator.is_valid(), "Parameter separator must be valid");
+    cudf::string_view d_separator( separator.data(), separator.size() );
```
Comment (vuule) [inline cpp/src/text/generate_ngrams.cu]:
```suggestion
    cudf::string_view const d_separator( separator.data(), separator.size() );
```

File: cpp/src/text/generate_ngrams.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/text/generate_ngrams.cu]: Apparently this is a common scenario when using Dask and we were told not to throw exceptions for empty columns.

File: cpp/src/text/generate_ngrams.cu (hunk 4 above)
Comment (davidwendt) [reply cpp/src/text/generate_ngrams.cu]: This is a mutable column data pointer and should not be const.

Comment (vuule) [reply cpp/src/text/generate_ngrams.cu]: From what I see, the data is mutable, but the pointer itself could be const. The data does not change the address.

File: cpp/src/text/generate_ngrams.cu (hunk 2 above)
Comment (vuule) [reply cpp/src/text/generate_ngrams.cu]: > Apparently this is a common scenario when using Dask and we were told not to throw exceptions for empty columns.

What about the other option, returning an empty column when there aren't enough elements to form a single ngram?

File: cpp/src/text/generate_ngrams.cu (hunk 4 above)
Comment (davidwendt) [reply cpp/src/text/generate_ngrams.cu]: I see what you mean. I that case, I'd rather remove the auto to make it clearer and use this instead.
```
char* const d_chars = chars_column->mutable_view().data<char>();
```

File: cpp/src/text/generate_ngrams.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/text/generate_ngrams.cu]: I think your first statement makes more sense. Here an empty column was indeed not passed but only _became empty_ by the filter and so therefore we should be allowed to throw an exception as per the ngram count check. This is cleaner and does not violate the empty-column rule.

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

**Pull request merged**